sql/src/main/java/io/crate/metadata/Signature.java

```diff
@@ -22,137 +22,363 @@
-                builder.add(new Signature(leftType, rightType));
-            }
-        }
-        return builder.build();
+    public static SignatureOperator ofIterable(Iterable<ArgMatcher> matchers) {
+        return new IterableArgs(matchers);
     }
 
-    private final ImmutableList<DataType> list;
-
     /**
-     * Defines which data types can be repeated unlimited to match this signature.
-     * The value will be used to build a subList of data types which will be used as a repeatable sequence.
-     *
-     * Example:
+     * Creates an operator which matches if the given positional argument matchers match the signature.
      *
-     *  list =                      [string, string, long]
-     *  varArgsStartingPosition =   1
+     * @param matchers positional argument matchers
+     * @return a new signature operator
+     */
+    public static SignatureOperator of(ArgMatcher... matchers) {
+        if (matchers.length == 0) {
+            return EMPTY;
+        }
+        return new VarArgs(matchers);
+    }
+
+    /**
+     * Returns an operator which matches and returns its input if the number of arguments in a signature are in a
+     * specific range.
      *
-     *     -> variable sequence =   [string, long]
+     * @param minSize the minimum number of arguments the signature is required to have.
+     * @param maxSize the maximum number of arguments the signature is required to have.
+     * @return an new signature operator
+     */
+    public static SignatureOperator numArgs(int minSize, int maxSize) {
+        Preconditions.checkArgument(minSize <= maxSize, "minSize needs to be smaller than maxSize");
+        return dataTypes -> (dataTypes.size() >= minSize && dataTypes.size() <= maxSize) ? dataTypes : null;
+    }
+
+    /**
+     * Returns an operator which matches and returns its input if the number of arguments matches.
      *
-     * Matching lists:
+     * @param numArgs the number of arguments the signature is required to have.
+     * @return an new signature operator
+     */
+    public static SignatureOperator numArgs(int numArgs) {
+        Preconditions.checkArgument(numArgs >= 0, "numArgs must not be negative");
+        if (numArgs == 0) {
+            return EMPTY;
+        }
+        return in -> in.size() == numArgs ? in : null;
+    }
+
+    /**
+     * Returns an operator with the given list of matchers and optional varArgs semantics.
      *
-     *  [string, string, long]
-     *  [string, string, long, string, long]
-     *  ...
+     * @param hasVarArgs        if true the final argument matcher is allowed to be repeated.
+     * @param strictVarArgTypes if true the matcher validates that varArgs are of the same type and also
+     *                          normalizes undefined types if possible.
+     * @param matchers          a new signature operator.
+     */
+    public static SignatureOperator of(boolean hasVarArgs, boolean strictVarArgTypes, ArgMatcher... matchers) {
```

Comment:
All callees currently call this always with `hasVarArgs=true` - Would it make sense to split this into two overloads:  `withStrictVarArgs(ArgMatcher...)` and `withLenientVarArgs`.

This would make the calls more readable than just `Signature.of(true, true, ...)`